stored_requests/events/http/http.go

```diff
@@ -0,0 +1,173 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	httpCore "net/http"
+	"time"
+
+	"golang.org/x/net/context/ctxhttp"
+
+	"github.com/buger/jsonparser"
+	"github.com/prebid/prebid-server/stored_requests/events"
+
+	"github.com/golang/glog"
+)
+
+// NewHTTPEvents makes an EventProducer which creates events by pinging an external HTTP API
+// for updates periodically. If refreshRate is negative, then the data will never be refreshed.
+//
+// It expects the following endpoint to exist remotely:
+//
+// GET {endpoint}
+//   -- Returns all the known Stored Requests and Stored Imps.
+// GET {endpoint}?last-modified={timestamp}
+//   -- Returns the Stored Requests and Stored Imps which have been updated since the last timestamp.
+//      This timestamp will be sent in the rfc3339 format, using UTC and no timezone shift.
+//      For more info, see: https://tools.ietf.org/html/rfc3339
+//
+// The responses should be JSON like this:
+//
+// {
+//   "requests": {
+//     "request1": { ... stored request data ... },
+//     "request2": { ... stored request data ... },
+//     "request3": { ... stored request data ... },
+//   },
+//   "imps": {
+//     "imp1": { ... stored data for imp1 ... },
+//     "imp2": { ... stored data for imp2 ... },
+//   }
+// }
+//
+// To signal deletions, the endpoint may return { "deleted": true }
+// in place of the Stored Data if the "last-modified" param existed.
+//
+func NewHTTPEvents(client *httpCore.Client, endpoint string, ctxProducer func() (ctx context.Context, canceller func()), refreshRate time.Duration) *httpEvents {
+	// If we're not given a function to produce Contexts, use the Background one.
+	if ctxProducer == nil {
+		ctxProducer = func() (ctx context.Context, canceller func()) {
+			return context.Background(), func() {}
+		}
+	}
+	e := &httpEvents{
+		client:        client,
+		ctxProducer:   ctxProducer,
+		endpoint:      endpoint,
+		lastUpdate:    time.Now().UTC(),
+		updates:       make(chan events.Save, 1),
+		invalidations: make(chan events.Invalidation, 1),
+	}
+	glog.Infof("Loading HTTP cache from GET %s", endpoint)
+	e.fetchAll()
+
+	go e.refresh(time.Tick(refreshRate))
+	return e
+}
+
+type httpEvents struct {
+	client        *httpCore.Client
+	ctxProducer   func() (ctx context.Context, canceller func())
+	endpoint      string
+	invalidations chan events.Invalidation
+	lastUpdate    time.Time
+	updates       chan events.Save
```

Comment:
If we're removing updates in favor of saves there are a few instances in this file where `updates` is referenced where it's actually a save.

Reply:
good call, thanks... updated now.